### PR TITLE
Introducing __clove_path_to_os function; Enhancements for __clove_path_basepath

### LINKED ISSUE #62: Double allocation in __clove_path_basepath
`__clove_path_basepath` uses strdup and calloc. This results in two separate memory allocations, which is inefficient and may cause memory fragmentation. A better approach would be to use only one allocation as mentioned in the function itself.
See: [clove-unit.h#L818](https://github.com/fdefelici/clove-unit/blob/d43981b346bfe374cbff6958092daa72aa2611c4/clove-unit.h#L818)

### PR DESCRIPTION
**Pull request by vonhoff**
Fix #62 
- Added unit tests based on the previous implementation to ensure original functionality.
- Removed usage of `__clove_string_strdup` in `__clove_path_basepath`.

### CODE REVIEW

**Comment by fdefelici** [comment]
The allocation has been reduced to one, but probably we loose clarity with a more algorithm approach.

Considering that the function is not critical for performance I'd prefer an approach that favor code readability (and then if with just one alloc, better :))

**Comment by vonhoff** [comment]
Another solution I had in mind would be the following, I have tried to make it a bit clearer and more readable. It may still be algorithmic, but I'm out of ideas. I will close this PR if a solution is not necessary. 
```
char* __clove_path_basepath(const char* a_path) {
    // Find the last path separator character in the input path.
    const char* last_char = a_path + strlen(a_path) - 1U;
    while (last_char > a_path && *last_char != '/' && *last_char != '\\') {
        --last_char;
    }

    // If there are no separators in the path, return the current directory path.
    if (last_char == a_path) {
        static char dot_path[3] = { '.', __CLOVE_PATH_SEPARATOR, '\0' };
        return strdup(dot_path);
    }

    size_t base_length = last_char - a_path;
    char* base_path = __CLOVE_MEMORY_CALLOC_TYPE_N(char, base_length + 1);
    memcpy(base_path, a_path, base_length);

    // Replace any '/' or '\\' characters in the base path with the appropriate path separator.
    for (size_t i = 0; i < base_length; i++) {
        if (base_path[i] == '/' || base_path[i] == '\\') {
            base_path[i] = __CLOVE_PATH_SEPARATOR;
        }
    }

    return base_path;
}
```

**Comment by fdefelici** [comment]
Second solution seems better.

I suggest to use some clove apis:
- __clove_memory_memcpy
- __clove_string_replace_char
- __clove_string_length

Note: Eventually we could wrap the two calls to `__clove_string_replace_char` into a new function `__clove_path_to_os(char*)`. 
We could also use this new function instead of #ifdef in unit test to obtain expected paths for assertion.

**Comment by vonhoff** [comment]
Apologies for the delay; I was on a break. I've now revised the code, including relevant unit tests. To test the new `__clove_path_to_os(char*)` function, I've relocated the #ifdef directives to these tests, ensuring that the function's behavior aligns with the operating system it runs on.

**Review by fdefelici — CHANGES_REQUESTED** [review]
I was imaging __clove_path_to_os like this: `void __clove_path_to_os(char*)`
to avoid allocation. Then in future if we need to keep the original string from modification we could implement a dedicated function.

File: clove-unit.h [hunk 1]
```diff
@@ -815,33 +816,34 @@ bool __clove_path_is_relative(const char* path) {
...
+    __clove_string_replace_char(os_path, '/', __CLOVE_PATH_SEPARATOR);
+    __clove_string_replace_char(os_path, '\\', __CLOVE_PATH_SEPARATOR);
+    return os_path;
+}
 
-    const char* last_addr = strrchr((const char*)path, __CLOVE_PATH_SEPARATOR);
-    int bytes_count;
+char* __clove_path_basepath(const char* a_path) {
+    // Find the last path separator character in the input path.
+    const char* last_char = a_path + strlen(a_path) - 1U;
```
Comment (fdefelici) [inline clove-unit.h]: here we could use `__clove_string_length`  instead of native strlen

File: clove-unit.h [hunk 2]
```diff
@@ -67,6 +67,7 @@ __CLOVE_EXTERN_C const char* __clove_get_exec_path();
 __CLOVE_EXTERN_C char* __clove_path_concat(const char separator, const char* path1, const char* path2);
 __CLOVE_EXTERN_C char* __clove_path_rel_to_abs_exec_path(const char* rel_path);
 __CLOVE_EXTERN_C bool __clove_path_is_relative(const char* path);
+__CLOVE_EXTERN_C char* __clove_path_to_os(const char* path);
```
Comment (fdefelici) [inline clove-unit.h]: for now a `void __clove_path_to_os(char*)` could do the job, avoiding allocation

**Review by fdefelici — CHANGES_REQUESTED** [review]
I was imaging __clove_path_to_os like this: void __clove_path_to_os(char*)
to avoid allocation. Then in future if we need to keep the original string from modification we could implement a dedicated function.

**Comment by vonhoff** [comment]
I applied the requested changes

**Review by fdefelici — APPROVED** [review]

**Pull request merged**